Add examples to [depr.volatile] (#3123)
cplusplus/draft: source/future.tex

prefix \tcode{++} and \tcode{\dcr} expressions\iref{expr.pre.incr}
of \tcode{volatile}-qualified arithmetic and pointer types are deprecated.

\begin{example}
\begin{codeblock}
volatile int velociraptor;
++velociraptor;                 // deprecated: is this a single memory operation, two, or a six-foot turkey?
\end{codeblock}
\end{example}

\pnum
Certain assignments
where the left operand is a \tcode{volatile}-qualified non-class type
are deprecated; see~\ref{expr.ass}.

\begin{example}
\begin{codeblock}
int neck, tail;
volatile int brachiosaur;

> tail = brachiosaur = neck;      // deprecated: does this touch brachiosaur once or twice?

Reviewer: ```suggestion
tail = brachiosaur = neck;      // deprecated: does this access brachiosaur once or twice?
```

Reviewer: It's not clear from this example which of the two assignment expressions in this statement is deprecated.

Reviewer: Could we have a `volatile_lhs += rhs` example in here as well?

\end{codeblock}
\end{example}

\pnum
A function type\iref{dcl.fct}
with a parameter with \tcode{volatile}-qualified type or
with a \tcode{volatile}-qualified return type is deprecated.

\begin{example}
\begin{codeblock}

> // Deprecated: a volatile return type has no meaning to the caller.

Reviewer: Capitalization inconsistent with the other examples.
```suggestion
// deprecated: a volatile return type has no meaning to the caller.
```

volatile amber jurassic();

> // The volatile left_arm and right_arm parameter are useless to the caller.

Reviewer: All by-value parameters are "useless to the caller". This doesn't tell me what's special about `volatile` by-value parameters that they should be deprecated.

void trex(volatile short left_arm, volatile short right_arm);
\end{codeblock}
\end{example}

\pnum
A structured binding\iref{dcl.struct.bind} of a \tcode{volatile}-qualified type
is deprecated.

\begin{example}
\begin{codeblock}
struct linhenykus { volatile short forelimb; };
void (linhenykus alvarezsauroid) {

>   auto& [what_is_this] = alvarezsauroid; // deprecated: very confusing

Reviewer: This should maybe have a more specific description than "very confusing". If we're going to start deprecating C++ constructs that are "very confusing", we're going to have a *ton* of work to do.

prefix \tcode{++} and \tcode{\dcr} expressions\iref{expr.pre.incr}
of \tcode{volatile}-qualified arithmetic and pointer types are deprecated.

\begin{example}
\begin{codeblock}
volatile int velociraptor;

> ++velociraptor;                 // deprecated: is this a single memory operation, or two

Reviewer 2: Please remove the attempt to give implicit rationale here. Just say that this case is deprecated, nothing more.

Author: Done.

> ++velociraptor;                 // deprecated: is this a single memory operation, two, or a six-foot turkey?

Reviewer 2: The tone of this comment is too flippant, and doesn't really work as a rationale for deprecation (since the same question applies for all non-deprecated volatile operations, which are all just as implementation-defined as each other).

Just saying "// deprecated" seems fine (here and below). We don't need to justify why things are deprecated.

Author: It's a [very Cultured reference](https://www.youtube.com/watch?v=TbrsQNeEyIg), but I can drop it if you want.

Reviewer 3: Quote from https://isocpp.org/std/standing-documents/sd-4-wg21-practices-and-procedures#code-of-conduct:
"Avoid metaphors, irony & be aware that jokes and humor may not translate"

Author: I dropped the reference, but I don't think we want to drop the explanations. I was explicitly asked for them in CWG.

\end{codeblock}
\end{example}

\pnum
Certain assignments
where the left operand is a \tcode{volatile}-qualified non-class type
are deprecated; see~\ref{expr.ass}.

\begin{example}
\begin{codeblock}
int neck, tail;
volatile int brachiosaur;
brachiosaur = neck;             // OK, a volatile store
tail = brachiosaur;             // OK, a volatile load

> tail = brachiosaur = neck;      // deprecated: does this access brachiosaur once or twice?

Reviewer 4: ```suggestion
tail = brachiosaur = neck;      // deprecated: does this access \tcode{brachiosaur} once or twice?
```

> brachiosaur += neck;            // deprecated: does this access brachiosaur once or twice?

Reviewer 4: ```suggestion
brachiosaur += neck;            // deprecated: does this access \tcode{brachiosaur} once or twice?
```

> brachiosaur = brachiosaur + neck; // OK, a volatile load, an addition, a volatile store

Reviewer 4: These comments need to be aligned at a multiple of 4, among other things. See https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#code-blocks.

\end{codeblock}
\end{example}

\pnum
A function type\iref{dcl.fct}
with a parameter with \tcode{volatile}-qualified type or
with a \tcode{volatile}-qualified return type is deprecated.

\begin{example}
\begin{codeblock}

> volatile amber jurassic(); // deprecated: a volatile return type has no meaning.

Reviewer 4: Comments above this one don't end in a full stop. I suggest looking around where this is added and following their ways.

> volatile amber jurassic();  // deprecated: a volatile return type has no meaning

Reviewer 3: The identifier "amber" does not appear to be defined.

// deprecated: volatile parameters aren't meaningful to the caller, volatile only applies within the function
void trex(volatile short left_arm, volatile short right_arm);

> void fly(volatile pterosaur* pteranodon);   // OK: the pointer isn't volatile, the data it points to is

Reviewer 3: Please add at least a declaration of "pterosaur", otherwise there's an undefined identifier used in this example.

> void fly(volatile struct pterosaur* pteranodon);   // OK: the pointer isn't volatile, the data it points to is

Reviewer: Why is this "OK:" when elsewhere you use "Ok, "?

Author: For consistency with the rest of the standard, where we do `OK:` 257 times and `OK,` 131 times ;-)

I've updated to be consistent here.

\end{codeblock}
\end{example}

\pnum
A structured binding\iref{dcl.struct.bind} of a \tcode{volatile}-qualified type
is deprecated.

\begin{example}
\begin{codeblock}
struct linhenykus { volatile short forelimb; };

> void (linhenykus alvarezsauroid) {

Reviewer 3: Function name is missing.

>   auto& [what_is_this] = alvarezsauroid; // deprecated: does the binding copy the forelimbs?

Reviewer 4: ```suggestion
  auto& [what_is_this] = alvarezsauroid; // deprecated: does the binding copy the \tcode{forelimb}s?
```

>   auto [what_is_this] = alvarezsauroid;    // deprecated: does the binding copy the \tcode{forelimb}s?

Reviewer 2: This example is not deprecated. The deprecated case is
```
struct whatever { int n; };
volatile auto [what_is_this] = whatever{1};
```

Author: Done.

>   auto& [what_is_this] = alvarezsauroid;    // deprecated: does the binding copy the \tcode{forelimb}s?

Reviewer 3: This is using "auto&", why would it copy anything at all?

\begin{codeblock}
int neck, tail;
volatile int brachiosaur;

> brachiosaur = neck;             // OK, a volatile store

Reviewer: "Ok" is an abbreviation of "Okay", not an acronym, so the "k" shouldn't be capitalized unless you intend the postal abbreviation for Oklahoma. Should these be "ok:" for consistency with "deprecated:"?

Author: "OK" is used 597 in the standard, and "Ok" 0 times :)

Reviewer 3: Deviating from the "OK" convention is not in scope for this particular pull request. Please create a new editorial issue if you feel discussion is warranted.

Author: OK.

Reviewer 5: "OK" is correct, it's not an abbreviation of okay. The "okay" spelling is a bacronym. "Ok" would be wrong.

Reviewer: I am willing to allow that "OK" is as correct as "Ok" - style guides vary, and the word has no clear etymology. Let's consider this point resolved. I'm Ok with "OK".

Reviewer: ...although I do find the colon vs. comma inconsistency in "deprecated: " and "OK, " disturbing.

\begin{example}
\begin{codeblock}
volatile amber jurassic(); // deprecated: a volatile return type has no meaning.

> // deprecated: volatile parameter aren't meaningful to the caller, volatile only applies within the function.

Reviewer: parameters
